WordPress/Jetpack/Classes/NUX/New Landing Screen/Views/InfiniteScrollerView.swift

```diff
@@ -0,0 +1,114 @@
+    private var singleViewHeight: CGFloat? {
+        guard let stackView = stackView, !stackView.arrangedSubviews.isEmpty else {
+            return nil
+        }
+
+        return stackView.arrangedSubviews[0].frame.size.height
+    }
+
+    /// Initializes a view that produces an infinite scrolling effect.
+    /// - Parameter viewBuilder: Closure that generates a `UIView`. Repeatedly called to fill a `UIStackView` that will be scrolled.
+    convenience init(_ viewBuilder: @escaping ((Int) -> UIView)) {
+        self.init()
+        self.viewBuilder = viewBuilder
+        displayLink = CADisplayLink(target: self, selector: #selector(step))
+
+        setupScrollView()
+        let stackView = setupStackView()
+        addSubview(stackView)
+
+        stackView.topAnchor.constraint(equalTo: topAnchor).isActive = true
+        stackView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
+        stackView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
+        stackView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
+        stackView.widthAnchor.constraint(equalTo: widthAnchor).isActive = true
+
+        self.stackView = stackView
+        addViewsToStackView()
+
+        displayLink?.add(to: .current, forMode: .default)
+    }
+
+    private func addViewsToStackView() {
+        guard let viewBuilder = viewBuilder else { return }
+        guard let stackView = stackView else { return }
+
+        /// Note: This class makes the assumption that the view generated by `viewBuilder()` is taller than
+        /// the scroll view's frame. To support views shorter than the frame it'd need to calculate how many times
+        /// to repeat it based on the built view's height.
+        ///
+        /// Currently, five views are generated because of an even / odd congruency requirement by the JetpackLandingScreen.
+        for i in 0...4 {
+            stackView.addArrangedSubview(viewBuilder(i))
+        }
+    }
+
+    private func setupScrollView() {
+        delegate = self
+        bounces = false
+        showsVerticalScrollIndicator = false
+        showsHorizontalScrollIndicator = false
+        isUserInteractionEnabled = false
+    }
+
+    private func setupStackView() -> UIStackView {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.distribution = .fillEqually
+        stackView.spacing = 0
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+
+        return stackView
+    }
+
+    /// Called on each `CADisplayLink` frame and sets the vertical content offset if a scroller delegate is set.
+    @objc private func step() {
+        guard let scrollerDelegate = self.scrollerDelegate else { return }
+        guard let singleViewHeight = self.singleViewHeight else { return }
+        // the rate shouldn't be higher than the size of the view
+        guard scrollerDelegate.rate < singleViewHeight else { return }
```

Comment:
same consideration here, as above, regarding the style

Reply:
Also, I think we can group these guard statements in one guard statement.

Reply:
Tweaked in 8df8479319c277a0bf3a72df836118ec54c0024b. 👍